Mandiant Advantage Integration initial implementation

miguelhdezsob opened the pull request:
## Summary of PR

* Generated the skeleton of the Mandiant Advantage integration package.
* Added data streams.
* Added data collection logic for all the data streams.
* Added the ingest pipeline for all the data streams.
* Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
* Added dashboards and visualizations.
* Added test for pipeline for all the data streams.
* Added system test cases for all the data streams.


## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [x] Update PR number in changelog once created

## How to test this PR locally

Pull the code locally and run `elastic-package test`. You might also want to build and manually install the integration if needed.

## Related issues

- N/A

## Screenshots

![image](https://github.com/elastic/integrations/assets/6812939/e635d2c1-11a5-47fb-b85d-043ea09114a0)
![image](https://github.com/elastic/integrations/assets/6812939/ee6d8e1d-033b-4da9-ba30-c3011b07260b)
![image](https://github.com/elastic/integrations/assets/6812939/3db11b85-1f9d-4012-a5ff-eef2225b18c3)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8390/detail/PR-8390/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8390/detail/PR-8390/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8390/detail/PR-8390/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8390/detail/PR-8390/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-11-02T21:41:40.834Z&rangeTo=2023-11-02T22:01:40.834Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=b6758e540079c8e18b61b04eeb6a97d9&transactionId=2b8535ada2e9bc7c)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-11-02T21:51:40.834+0000


* Duration: 3 min 12 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8390/detail/PR-8390/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8390/runs/1/steps/81/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/integrations.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8390/runs/1/steps/89/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/miguelhdezsob return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/miguelhdezsob : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/miguelhdezsob : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/orgs/members#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

kgeller commented on packages/ti_mandiant_advantage/_dev/build/build.yml:
```diff
@@ -0,0 +1,3 @@
+dependencies:
+  ecs:
+    reference: git@8.7
```
Could we update this to the latest? After today it'll be 8.11

kgeller commented on packages/ti_mandiant_advantage/manifest.yml:
```diff
@@ -0,0 +1,50 @@
+format_version: 2.5.1
```
Could we upgrade this to 3.0.0? That is the new standard version for our integrations

kgeller commented on packages/ti_mandiant_advantage/manifest.yml:
```diff
@@ -0,0 +1,50 @@
+            required: false
+            show_user: true
+            description: The request tracer logs requests and responses to the agent's local file-system for debugging configurations. Enabling this request tracing compromises security and should only be used for debugging. See [documentation](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html#_request_tracer_filename) for details.
+owner:
+  github: elastic/security-external-integrations
```
```suggestion
owner:
  github: elastic/security-external-integrations
  type: partner
```

kgeller commented:
/test

kgeller commented:
```
Error: there is no owner for "packages/ti_mandiant_advantage" in ".github/CODEOWNERS"
--
  | 🚨 Error: The command exited with status 1
```

Currently the build is failing because there's no entry in [codeowners](https://github.com/elastic/integrations/blob/main/.github/CODEOWNERS). Once that gets added, we should get a 'real' run of CI

miguelhdezsob commented:
> ```
> Error: there is no owner for "packages/ti_mandiant_advantage" in ".github/CODEOWNERS"
> --
>   | 🚨 Error: The command exited with status 1
> ```
> 
> Currently the build is failing because there's no entry in [codeowners](https://github.com/elastic/integrations/blob/main/.github/CODEOWNERS). Once that gets added, we should get a 'real' run of CI

Thanks for the review! I believe I have addressed your comments and added the CODEOWNERS entry. This should be ready for another review.

kgeller commented:
/test

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`1/1`) | :green_heart: 13.636
Classes | 100.0% (`1/1`) | :green_heart: 13.636
Methods | 93.333% (`14/15`) | :+1: 10.0
Lines | 97.742% (`303/310`) | :+1: 12.721
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

kgeller commented on packages/ti_mandiant_advantage/data_stream/threat_intelligence/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,399 @@
+      ignore_missing: true
+- set:
+      field: ecs.version
+      value: 8.7.0
```
```suggestion
      value: 8.11.0
```

kgeller commented on packages/ti_mandiant_advantage/data_stream/threat_intelligence/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,399 @@
+---
+description: Pipeline for processing Mandiant IOCs.
+processors:
+- rename:
+      field: message
+      target_field: event.original
+      ignore_missing: true
```
```suggestion
- rename:
      field: message
      target_field: event.original
      ignore_missing: true
      if: 'ctx.event?.original == null'
```

This if check just helps us prevent conflict in the event the document already has the field set ie when sending from Logstash

kgeller commented:
@miguelhdezsob CI is looking good. Just a couple final comments and we should be good to go :)

miguelhdezsob commented on packages/ti_mandiant_advantage/data_stream/threat_intelligence/agent/stream/httpjson.yml.hbs:
```diff
@@ -0,0 +1,60 @@
+  - delete:
+      target: header.User-Agent
+  - set:
+      # NOTE: The "Build" version must be kept in sync with this package's version.
+      target: header.User-Agent
+      value: '[[userAgent "(Elastic Integration; Mandiant Advantage; Build/0.1.0)"]]'
+  - set:
+      # NOTE: The version must be kept in sync with this package's version.
+      target: header.X-App-Name
+      value: '[["MA-Elastic-Integration-0.1.0)"]]'
```
It is not a hard requirement, but we do have a best practice of including   the version in this header for any of our integrations. This is used for usage tracking and troubleshooting on our end.

I did leave a comment saying this needs to be manually kept in sync, but could remove the version if you insist.

kgeller replied:
We can leave it as is. I found out that we do this in a couple of other places as well.

kgeller commented:
/test

kgeller reviewed (approved):
LGTM

elasticmachine commented:
Package ti_mandiant_advantage - 0.1.0 containing this change is available at https://epr.elastic.co/search?package=ti_mandiant_advantage
